### PR TITLE
[SYCL] Test corrections after moving bfloat16 support out of experimental status.

### PR DESCRIPTION
**Pull request by rdeodhar**
Signed-off-by: Rajiv Deodhar <rajiv.deodhar@intel.com>

### CODE REVIEW

**Review by kbobrovs — DISMISSED** [review]
ESIMD LGTM

**Review by v-klochkov — DISMISSED** [review]

**Comment by gmlueck** [comment]
It would be good to have some tests here for the various conversions between `bfloat16`, `float`, and `sycl::half`.

File: SYCL/BFloat16/bfloat16_type.cpp
Comment (gmlueck) [inline SYCL/BFloat16/bfloat16_type.cpp]: It would be nice to say something about AOT here since the limitation is specific to AOT.  Don't you also have a problem on GPU when the GPU does not have native support?

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_type.cpp]: This is a test, and it is not doing AOT, so why bring that up?

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_type.cpp]: When we have compiler driver support for AOT everywhere, this test can be extended to use both JIT and AOT.

Comment (gmlueck) [reply SYCL/BFloat16/bfloat16_type.cpp]: Oh, I was confused about the test.  I thought it was doing AOT.  If it is a JIT mode test, why doesn't CPU work?

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_type.cpp]: Need to look into it. CPU should use the emulation method always. I thought the test failed because not all CPUs support bfloat16.

Comment (gmlueck) [reply SYCL/BFloat16/bfloat16_type.cpp]: OK, so we'll fix the problem and re-enable the test before merging?

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_type.cpp]: Building for the _CPU device_ is essentially an AOT build. Currently, AOT builds always pull in the native bfloat16 library, which isn't supported on the CPU.
We will have to defer _CPU device_ support for bfloat16 until the driver is changed to fully support AOT. The driver will select native bfloat16 library only for Xe GPUs and fallback bfloat16 library for other GPUs and the CPU.

Comment (gmlueck) [reply SYCL/BFloat16/bfloat16_type.cpp]: Something doesn't sound right.  You can definitely run on the CPU device without AOT compiling.  For example:

```
$ clang++ -fsycl -o very-simple very-simple.cpp    # Compiles without AOT
$ SYCL_DEVICE_FILTER=level_zero:gpu ./very-simple  # Runs on GPU
$ SYCL_DEVICE_FILTER=opencl:cpu ./very-simple      # Runs on CPU
```

Is there something weird about `%CPU_RUN_PLACEHOLDER` that forces an AOT compile?

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_type.cpp]: _CPU device_ support is not completely broken, but just for bfloat16.
This is not a regression. Even earlier, CPU device did not work with bfloat16.

When doing JIT compilation for CPU there is no target device check. The same libraries are always linked in. Currently for bfloat16 this is the native library (so that AOT for PVC works). When AOT is fixed to check which device is being compiled for, and the appropriate bfloat16 library is linked in, that will also correct the JIT case for CPU.

Comment (gmlueck) [reply SYCL/BFloat16/bfloat16_type.cpp]: > When doing JIT compilation for CPU there is no target device check

There is no such thing as JIT compilation for the CPU.  The JIT compilation mode just generates SPIR-V, and the target device isn't known until runtime.  It sounds like there might be a bug in the runtime code that decides which of the bfloat16 spirv files to link at runtime (the native version vs. the fallback version).  I wonder if that logic is somehow broken for the CPU device or for the OpenCL backend?

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_type.cpp]: I have made some temporary changes in the compiler driver to support AOT.

**Comment by rdeodhar** [comment]
This PR requires corresponding changes in intel/llvm, made in PR 6524. That has been successful.

File: SYCL/BFloat16/bfloat16_type.cpp
Comment (gmlueck) [reply SYCL/BFloat16/bfloat16_type.cpp]: Cool!  Does that mean we can re-enable testing on CPU?  BTW, shouldn't there also be a `GPU_RUN_PLACEHOLDER` here too?

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_type.cpp]: Yes, I've now enabled testing on CPU and GPU.

Comment (gmlueck) [reply SYCL/BFloat16/bfloat16_type.cpp]: Nice!

File: SYCL/BFloat16/bfloat16_type.cpp
Comment (JackAKirk) [inline SYCL/BFloat16/bfloat16_type.cpp]: Currently this will only check the software emulation cases are compiled and run correctly for the CUDA backend case. You can at least check the native versions are also compiling correctly if you add:
```suggestion
// RUN: %if cuda %{%clangxx -fsycl -fsycl-targets=%sycl_triple -Xsycl-target-backend --cuda-gpu-arch=sm_80 %s -o %t.out %}
// TODO enable the below when CI supports >=sm_80
// RUNx: %if cuda %{%GPU_RUN_PLACEHOLDER %t.out %}
// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
```

Also I think you could now delete the bfloat16_type_cuda.cpp test since it is really just duplicating this test.

Comment (gmlueck) [reply SYCL/BFloat16/bfloat16_type.cpp]: Does the test pool contain any systems with CUDA devices that are >= sm_80?  If so, can we write a separate test that uses REQUIRE to enable a test only on that system?  This would allow us to compile and run the test with native support.

Comment (JackAKirk) [reply SYCL/BFloat16/bfloat16_type.cpp]: Last I heard there was only sm75. Is there an advantage to having a separate test with REQUIRE for such as case? As opposed to just adding %if cases to an existing file? I'm just trying to understand the CI better.

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_type.cpp]: I've reenabled the CUDA execution option. I don't know if REQUIRE causes the test to be run on a pre-selected machine. Will find out.

File: SYCL/BFloat16/bfloat16_example.cpp
Comment (gmlueck) [inline SYCL/BFloat16/bfloat16_example.cpp:31]: I think we can add the following test cases:

```
// RUN: %clangxx -fsycl -fsycl-targets=spir64,spir64_gen -Xsycl-target-backend=spir64_gen "-device pvc" %s -o %t.out
// RUN: %CPU_RUN_PLACEHOLDER %t.out

// RUN: %clangxx -fsycl -fsycl-targets=spir64_x86_64,spir64_gen -Xsycl-target-backend=spir64_gen "-device pvc" %s -o %t.out
// RUN: %CPU_RUN_PLACEHOLDER %t.out
```

These are interesting because they compile for PVC (native bfloat support) and run on CPU (fallback support).

File: SYCL/BFloat16/bfloat16_example.cpp
Comment (gmlueck) [inline SYCL/BFloat16/bfloat16_example.cpp]: Can we check here that the answer is expected?

```
float result_host = foo(7.0f, 8.1f);
if (std::abs(15.1f - result_host) > 0.01) {
  /* test fails */
}
```

File: SYCL/BFloat16/bfloat16_example.cpp
Comment (gmlueck) [inline SYCL/BFloat16/bfloat16_example.cpp]: And here we can check that the result on the device is exactly the same as the result on the host:

```
float result_device = hostOutAcc[2];
if (result_device != result_host) {
  /* test fails */
}
```

File: SYCL/BFloat16/bfloat16_example.cpp
Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_example.cpp:31]: Done.

File: SYCL/BFloat16/bfloat16_example.cpp
Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_example.cpp]: Yes, done.

File: SYCL/BFloat16/bfloat16_example.cpp
Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_example.cpp]: Done.

**Comment by kbobrovs** [comment]
ESIMD tests fail with
llvm-test-suite/SYCL/ESIMD/esimd_test_utils.hpp:556:24: error: no member named 'bfloat16' in namespace 'sycl::ext::oneapi'
[2022-10-25T18:32:40.795Z] TID(sycl::ext::oneapi::bfloat16)

if proper header included?

**Comment by rdeodhar** [comment]
Only the results from "Jenkins/llvm-test-suite" count. That is a cross-test with intel/llvm PR 6524.
Other testing will fail because of mismatched headers, and should not be paid attention to. If you know of a way to avoid that testing, please let me know.

File: SYCL/BFloat16/bfloat16_example.cpp
Comment (gmlueck) [inline SYCL/BFloat16/bfloat16_example.cpp]: Why are we removing the AOT GPU tests?  Is it because the CI system might not have a gen9 GPU?  If that is the case, can we somehow conditionally run the test only if the CI system is gen9?

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_example.cpp]: There isn't a way to check/select a particular Intel GPU in the CI infrastructure. The scenarios in the test are covered by the clang/Driver lit tests which check that the proper libraries are used. I could add a check in the test to check GPU type, but narrowing the scope of runnable tests isn't going to necessarily increase coverage, if the expected behavior has been checked in other ways.

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_example.cpp]: Correction. There does appear to be a way to select a specific Intel GPU. Let me try using that feature.

**Comment by rdeodhar** [comment]
Cross testing with intel/llvm PR 6524 is successful. This and that PR need to be merged in together.

**Comment by pvchupin** [comment]
This PR requires review from all: @intel/dpcpp-esimd-reviewers , @intel/dpcpp-tools-reviewers , @intel/llvm-reviewers-runtime

**Review by kbobrovs — APPROVED** [review]
ESIMD tests LGTM

**Review by smaslov-intel — APPROVED** [review]
LGTM

File: SYCL/BFloat16/bfloat16_builtins.cpp
Comment (asudarsa) [inline SYCL/BFloat16/bfloat16_builtins.cpp:18]: In Line 225, should it be ext_oneapi_bfloat16_math_functions? 

Thanks

Comment (rdeodhar) [reply SYCL/BFloat16/bfloat16_builtins.cpp:18]: Yes, it should be ext_oneapi_bfloat16_math_functions. Made the change.

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks

**Review by v-klochkov — APPROVED** [review]

**Pull request merged**
